ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/GSM.hs

```diff
@@ -0,0 +1,698 @@
+    ExtendSelection{} ->
+        Unit
+    ModifyCandidate{} ->
+        Unit
+    NewCandidate{} ->
+        Unit
+    ReadJudgment ->
+        ReadThisJudgment $ j
+    ReadMarker ->
+        ReadThisMarker $ toMarker j
+    StartIdling{} ->
+        Unit
+    TimePasses{} ->
+        Unit
+  where
+    j = toJudgment $ mState model
+
+sm ::
+     IOLike m
+  => Vars m
+  -> LedgerStateJudgement
+  -> QSM.StateMachine Model Command m Response
+sm vars j = QSM.StateMachine {
+    QSM.cleanup = \_model -> pure ()
+  ,
+    QSM.generator = generator
+  ,
+    QSM.initModel = initModel j
+  ,
+    QSM.invariant = Nothing
+  ,
+    QSM.mock = mock
+  ,
+    QSM.postcondition = postcondition
+  ,
+    QSM.precondition = precondition
+  ,
+    QSM.semantics = semantics vars
+  ,
+    QSM.shrinker = shrinker
+  ,
+    QSM.transition = transition
+  }
+
+prop_sequential ::
+     LedgerStateJudgement
+  -> QC.Property
+prop_sequential j0 =
+    QSM.forAllCommands
+        (sm (undefined :: Vars IO) j0)   -- NB the specific IO type is unused here
+        mbMinimumCommandLen
+        predicate
+  where
+    mbMinimumCommandLen = Just 20
+
+    predicate :: QSM.Commands Command Response -> QC.Property
+    predicate cmds = runSimOrThrow $ do
+        varSelection  <- newTVarIO (mSelection $ initModel j0)
+        varCandidates <- newTVarIO Map.empty
+        varIdlers     <- newTVarIO Set.empty
+        varJudgment   <- newTVarIO j0
+        varMarker     <- newTVarIO (toMarker j0)
+        let vars =
+                Vars
+                    varSelection
+                    varCandidates
+                    varIdlers
+                    varJudgment
+                    varMarker
+        let sm' = sm vars j0
+        let gsm = GSM.realGsmEntryPoints GSM.GsmView {
+                GSM.antiThunderingHerd = Nothing
+              ,
+                GSM.candidateOverSelection = candidateOverSelection
+              ,
+                GSM.durationUntilTooOld = Just durationUntilTooOld
+              ,
+                GSM.equivalent = (==)   -- unsound, but harmless in this test
+              ,
+                GSM.getChainSyncCandidates = readTVar varCandidates
+              ,
+                GSM.getChainSyncIdlers = readTVar varIdlers
+              ,
+                GSM.getCurrentSelection = readTVar varSelection
+              ,
+                GSM.minCaughtUpDuration = thrashLimit
+              ,
+                GSM.setCaughtUpPersistentMark = \b ->
+                    atomically $ do
+                        writeTVar varMarker $ if b then Present else Absent
+              ,
+                GSM.varLedgerStateJudgement = varJudgment
+              }
+            gsmEntryPoint = case j0 of
+                TooOld      -> GSM.enterOnlyBootstrap gsm
+                YoungEnough -> GSM.enterCaughtUp      gsm
+
+        (hist, _model, res) <- id
+          $ withAsync gsmEntryPoint
+          $ \_async -> runCommands' (pure sm') cmds
```

Comment:
Thanks for the note. (Again, I actually needed this during a debugging session.)

I went with `Async.poll`, because I don't want to abort the main thread, since it's the one that ends by nicely rendering the repro commands etc.